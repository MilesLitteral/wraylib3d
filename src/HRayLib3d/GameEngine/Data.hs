module HRayLib3d.GameEngine.Data (
        -- module HRayLib3d.GameEngine.Data.BSP,
        -- module HRayLib3d.GameEngine.Data.GameCharacter,
        -- module HRayLib3d.GameEngine.Data.GLB,
        -- module HRayLib3d.GameEngine.Data.GLTF,
        -- module HRayLib3d.GameEngine.Data.Material,
        -- module HRayLib3d.GameEngine.Data.MD3,
        -- module HRayLib3d.GameEngine.Data.OBJ
    ) where

    -- import HRayLib3d.GameEngine.Data.BSP
    -- import HRayLib3d.GameEngine.Data.GameCharacter
    -- import HRayLib3d.GameEngine.Data.GLB
    -- import HRayLib3d.GameEngine.Data.GLTF
    -- import HRayLib3d.GameEngine.Data.Material
    -- import HRayLib3d.GameEngine.Data.MD3
    -- import HRayLib3d.GameEngine.Data.OBJ