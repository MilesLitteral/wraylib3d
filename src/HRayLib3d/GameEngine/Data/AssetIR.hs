module HRayLib3d.GameEngine.Data.AssetIR where