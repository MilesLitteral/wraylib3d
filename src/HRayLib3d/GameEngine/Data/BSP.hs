{-# LANGUAGE OverloadedStrings, DeriveGeneric, RecordWildCards  #-}

module HRayLib3d.GameEngine.Data.BSP where

import Text.XML.Writer ()

import GHC.Generics    (Generic)
import Data.Aeson      --(FromJSON(parseJSON), ToJSON(toJSON), Object, object, KeyValue((.=)))
import Data.Binary     ()
import Data.Word       (Word8)
import Data.ByteString (ByteString)
import Data.Vector     (Vector)
import Data.Vect       (Vec3(..), Vec2(..), Vec4(..) )
import Data.Vect.Float.Instances ()

{-
    Information:
        http://graphics.stanford.edu/~kekoa/q3/
        http://www.mralligator.com/q3/
-}

data Model
    = Model
    { mdMins         :: !Vec3
    , mdMaxs         :: !Vec3
    , mdFirstSurface :: !Int
    , mdNumSurfaces  :: !Int
    , mdFirstBrush   :: !Int
    , mdNumBrushes   :: !Int
    } deriving (Eq, Show, Generic)

data Shader
    = Shader
    { shName         :: !ByteString
    , shSurfaceFlags :: !Int
    , shContentFlags :: !Int
    } deriving (Eq, Show, Generic)

data Plane
    = Plane
    { plNormal :: !Vec3
    , plDist   :: !Float
    } deriving (Eq, Show, Generic)

data Node
    = Node
    { ndPlaneNum :: !Int
    , ndChildren :: !(Int,Int)
    , ndMins     :: !Vec3
    , ndMaxs     :: !Vec3
    } deriving (Eq, Show, Generic)

data Leaf
    = Leaf
    { lfCluster          :: !Int
    , lfArea             :: !Int
    , lfMins             :: !Vec3
    , lfMaxs             :: !Vec3
    , lfFirstLeafSurface :: !Int
    , lfNumLeafSurfaces  :: !Int
    , lfFirstLeafBrush   :: !Int
    , lfNumLeafBrushes   :: !Int
    } deriving (Eq, Show, Generic)

data BrushSide
    = BrushSide
    { bsPlaneNum  :: !Int
    , bsShaderNum :: !Int
    } deriving (Eq, Show)

data Brush
    = Brush
    { brFirstSide :: !Int
    , brNumSides  :: !Int
    , brShaderNum :: !Int
    } deriving (Eq, Show)

data Fog
    = Fog
    { fgName        :: !ByteString
    , fgBrushNum    :: !Int
    , fgVisibleSide :: !Int
    } deriving (Eq, Show, Generic)

data DrawVertex
    = DrawVertex
    { dvPosition    :: !Vec3
    , dvDiffuseUV   :: !Vec2
    , dvLightmaptUV :: !Vec2
    , dvNormal      :: !Vec3
    , dvColor       :: !Vec4
    } deriving (Eq, Show, Generic)

data SurfaceType
    = Planar
    | Patch
    | TriangleSoup
    | Flare
    deriving (Eq, Show, Enum, Generic)

data Surface
    = Surface
    { srShaderNum      :: !Int
    , srFogNum         :: !Int
    , srSurfaceType    :: !SurfaceType
    , srFirstVertex    :: !Int
    , srNumVertices    :: !Int
    , srFirstIndex     :: !Int
    , srNumIndices     :: !Int
    , srLightmapNum    :: !Int
    , srLightmapPos    :: !Vec2
    , srLightmapSize   :: !Vec2
    , srLightmapOrigin :: !Vec3
    , srLightmapVec1   :: !Vec3
    , srLightmapVec2   :: !Vec3
    , srLightmapVec3   :: !Vec3
    , srPatchSize      :: !(Int,Int)
    } deriving (Eq, Show, Generic)

newtype Lightmap = Lightmap { lmMap :: ByteString } deriving (Eq, Show, Generic)

data LightGrid   = LightGrid deriving (Eq, Show, Generic)

data Visibility
    = Visibility
    { vsNumVecs     :: !Int
    , vsSizeVecs    :: !Int
    , vsVecs        :: !(Vector Word8)
    } deriving (Eq, Show, Generic)

data BSPLevel
    = BSPLevel
    { blEntities     :: !ByteString
    , blShaders      :: !(Vector Shader)
    , blPlanes       :: !(Vector Plane)
    , blNodes        :: !(Vector Node)
    , blLeaves       :: !(Vector Leaf)
    , blLeafSurfaces :: !(Vector Int)
    , blLeafBrushes  :: !(Vector Int)
    , blModels       :: !(Vector Model)
    , blBrushes      :: !(Vector Brush)
    , blBrushSides   :: !(Vector BrushSide)
    , blDrawVertices :: !(Vector DrawVertex)
    , blDrawIndices  :: !(Vector Int)
    , blFogs         :: !(Vector Fog)
    , blSurfaces     :: !(Vector Surface)
    , blLightmaps    :: !(Vector Lightmap)
    , blLightgrid    :: !(Vector LightGrid)
    , blVisibility   :: !Visibility
    } deriving (Eq, Show, Generic)

instance ToJSON Vec2 where
  toJSON (Vec2 x y) = object ["vec2X" .= x, "vec2Y" .= y ]

instance ToJSON Vec3 where
  toJSON (Vec3 x y z) = object ["vec3X" .= x, "vec3Y" .= y, "vec3Z" .= z ]

instance ToJSON Vec4 where
    toJSON (Vec4 w x y z) = object ["vec4W" .= w, "vec4X" .= x, "vec4Y" .= y, "vec4Z" .= z ]

instance ToJSON Shader where
    toJSON Shader{..} = object [ "shName" .= show shName, "shSurfaceFlags" .= shSurfaceFlags, "shContentFlags" .= shContentFlags]

instance ToJSON  Fog where
    toJSON Fog{..} = object ["fgName" .= (show fgName), "fgBrushNum" .= fgBrushNum, "fgVisibleSide" .= fgVisibleSide ]

instance ToJSON SurfaceType
instance ToJSON  Surface
instance ToJSON  Brush where
    toJSON Brush{..} = object[ "brFirstSide" .= brFirstSide, "brNumSides" .= brNumSides, "brShaderNum" .= brShaderNum]

instance ToJSON  Model
instance ToJSON  Leaf
instance ToJSON  Node where
    toJSON Node{..} = object[ "ndPlaneNum" .= ndPlaneNum, "ndChildren" .= ndChildren, "ndMins" .= ndMins, "ndMaxs" .= ndMaxs]

instance ToJSON  Plane
instance ToJSON  Lightmap where
    toJSON Lightmap{..} = object [ "lmMap" .= show lmMap ]

instance ToJSON  BrushSide where
    toJSON BrushSide{..} = object ["bsPlaneNum" .= bsPlaneNum, "bsShaderNum" .= bsShaderNum]

instance ToJSON  LightGrid
instance ToJSON  DrawVertex
instance ToJSON  Visibility

instance ToJSON  BSPLevel where
    toJSON BSPLevel{..} = object [ 
          "entities"     .= show blEntities
        , "shaders"      .= blShaders
        , "planes"       .= blPlanes
        , "nodes"        .= blNodes
        , "leaves"       .= blLeaves
        , "leafSurfaces" .= blLeafSurfaces
        , "leafBrushes"  .= blLeafBrushes
        , "models"       .= blModels
        , "brushes"      .= blBrushes
        , "brushSides"   .= blBrushSides
        , "drawVertices" .= blDrawVertices
        , "drawIndices"  .= blDrawIndices
        , "fogs"         .= blFogs
        , "surfaces"     .= blSurfaces
        , "lightmaps"    .= blLightmaps
        , "lightgrid"    .= blLightgrid
        , "visibility"   .= blVisibility
        ] 

instance FromJSON  LightGrid
instance FromJSON  Visibility

-- instance FromJSON  ByteString
-- instance FromJSON  Vec2
-- instance FromJSON  Vec3
-- instance FromJSON  Vec4

-- instance FromJSON  Node
-- instance FromJSON  SurfaceType
-- instance FromJSON  Surface
-- instance FromJSON  Lightmap

-- instance FromJSON  BSPLevel --where             
    -- parseJSON (Object v) = 
    --     BSPLevel <$> v     .: "entities"      
    --                 <*>  v .: "shaders" 
    --                 <*>  v .: "planes"        
    --                 <*>  v .: "nodes"         
    --                 <*>  v .: "leaves"        
    --                 <*>  v .: "leafSurfaces"  
    --                 <*>  v .: "leafBrushes"   
    --                 <*>  v .: "models"        
    --                 <*>  v .: "brushes"       
    --                 <*>  v .: "brushSides"    
    --                 <*>  v .: "drawVertices"  
    --                 <*>  v .: "drawIndices"   
    --                 <*>  v .: "fogs"          
    --                 <*>  v .: "surfaces"      
    --                 <*>  v .: "lightmaps"     
    --                 <*>  v .: "lightgrid"     
    --                 <*>  v .: "visibility"  

-- --fix this later for JSON and XML Support
-- TODO: a custom data XML object with an ToXML and FromXML allegory
-- that should rival Data.Aeson's support of JSON, this is not a priority
-- but something todo

-- instance ToXML BSPLevel where
--     toXML bspLevel = element "BSPLevel" $ many "entry" $ show $ blEntities bspLevel

-- map (\x -> element "container" $ many . show "wrapper" . blEntities)
-- instance Binary XML
-- instance (GHC.Generics.Generic XML)