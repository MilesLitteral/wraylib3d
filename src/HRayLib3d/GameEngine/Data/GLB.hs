module HRayLib3d.GameEngine.Data.GLB where

import Data.Text
import Data.Map (Map)
import Codec.GLB
import Codec.GlTF.Material

-- No Animation, for Static Models or bundles of multiple individual asset
-- This could be a bundle (in Chunks) of the following:
-- GlTFs
-- Textures
-- Misc Data 

type GLBSkin     = Map Text Material
newtype GLBModel = GLBModel { rawGlb :: GLB } deriving Show
