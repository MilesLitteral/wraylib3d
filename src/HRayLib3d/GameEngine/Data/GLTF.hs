module HRayLib3d.GameEngine.Data.GLTF (
    Frame(..)
    ,Tag(..)
    ,Shader(..)
    ,Surface(..)
    ,GLTFModel(..)
    ,GLTFSkin
    ,GLTFAnimation
)where

import Data.Int
import Data.Map (Map)
import Data.HashMap.Strict
import Data.Vect hiding (Vector)
import Data.Vector (Vector)
import Data.ByteString (ByteString)
import qualified Data.Vector.Storable as SV

import Codec.GlTF
import Codec.GlTF.Texture
import Codec.GlTF.Material
import Codec.GlTF.Animation
import Codec.GlTF.Buffer
import Codec.GlTF.Accessor
-- GlTF	 
-- asset :: Asset	 
-- extensionsUsed :: Maybe (Vector Text)	 
-- extensionsRequired :: Maybe (Vector Text)	 
-- accessors :: Maybe (Vector Accessor)	 
-- animations :: Maybe (Vector Animation)	 
-- buffers :: Maybe (Vector Buffer)	 
-- bufferViews :: Maybe (Vector BufferView)	 
-- cameras :: Maybe (Vector Camera)	 
-- images :: Maybe (Vector Image)	 
-- materials :: Maybe (Vector Material)	 
-- meshes :: Maybe (Vector Mesh)	 
-- nodes :: Maybe (Vector Node)	 
-- samplers :: Maybe (Vector Sampler)	 
-- scenes :: Maybe (Vector Scene)	 
-- skins :: Maybe (Vector Skin)	 
-- textures :: Maybe (Vector Texture)	 
-- extensions :: Maybe Object	 
-- extras :: Maybe Value

data Frame
    = Frame
    { frGlTFMins    :: !Vec3
    , frGlTFMaxs    :: !Vec3
    , frGlTFOrigin  :: !Vec3
    , frGlTFRadius  :: !Float
    , frGlTFName    :: !ByteString
    } deriving Show

data Tag
    = Tag
    { tgGlTFName         :: !ByteString
    , tgGlTFOrigin       :: !Vec3
    , tgGlTFRotationMat  :: !Mat3
    } deriving Show

-- (One Day) Codec.GlTF Material -> Shader (Material?)
-- Shaders can be pulled out of GLTFs from:
-- GlTF.extensionsUsed  
-- GlTF.extensionsRequired
-- The necessary data looks like this:
-- {
--     "extensions": {
--         "KHR_techniques_webgl": {
--             "programs": [
--                 {
--                     "fragmentShader": 0,
--                     "vertexShader": 1
--                 }
--             ],
--             "shaders": [
--                 {
--                     "type": 35632,
--                     "uri": "duck0FS.glsl"
--                 },
--                 {
--                     "type": 35633,
--                     "uri": "duck0VS.glsl"
--                 }
--             ],
--             "techniques": [
--                 {
--                     "program": 0,
--                     "attributes": {
--                         "a_position": {
--                             "semantic": "POSITION"
--                         }
--                     },
--                     "uniforms": {
--                         "u_modelViewMatrix": {
--                             "type": 35676,
--                             "semantic": "MODELVIEW"
--                         }
--                     }
--                 }
--             ]
--         }
--     }
-- }

data Shader
    = Shader
    { shGlTFName    :: !ByteString
    , shGlTFIndex   :: !Int
    } deriving Show

data Surface
    = Surface
    { srGlTFName        :: !ByteString
    , srGlTFShaders     :: !(Vector Shader) -- apply a generic, applicable, glsl shader for now -- Materials come later
    , srGlTFTriangles   :: !(SV.Vector Int32)
    , srGlTFTexCoords   :: !(SV.Vector Vec2)
    , srGlTFXyzNormal   :: !(Vector (SV.Vector Vec3,SV.Vector Vec3))
    } deriving Show

data GLTFModel
    = GLTFModel
    { gltfFrames      :: !(Vector Frame) -- GlTF.accessors :: Maybe (Vector Accessor) -> gltfFrames :: !(Vector Frame)
    , gltfTags        :: !(Vector (HashMap ByteString Tag))
    , gltfSurfaces    :: !(Vector Surface)
    } deriving Show

type GLTFSkin      = Map String Texture
type GLTFAnimation = Map String Animation