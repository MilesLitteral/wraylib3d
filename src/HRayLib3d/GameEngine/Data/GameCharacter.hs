{-# LANGUAGE DeriveGeneric #-}
module HRayLib3d.GameEngine.Data.GameCharacter where

import GHC.Generics        ( Generic  )
import Data.Hashable       ( Hashable )
import Data.HashMap.Strict ( HashMap  )
import LambdaCube.Linear   (V3(..))

data Animation
  = Animation
  { aFirstFrame :: !Int
  , aNumFrames  :: !Int
  , aLoopFrames :: !Int  -- 0 to numFrames
  , aFrameLerp  :: !Int  -- msec between frames
  , aIntialLerp :: !Int  -- msec to get to first frame
  , aReversed   :: !Bool -- true if animation is reversed
  , aFlipFlop   :: !Bool -- true if animation should flipflop back to base
  }
  deriving (Show, Eq, Ord)

data FootStep
  = FOOTSTEP_NORMAL
  | FOOTSTEP_BOOT
  | FOOTSTEP_FLESH
  | FOOTSTEP_MECH
  | FOOTSTEP_ENERGY
  | FOOTSTEP_METAL
  | FOOTSTEP_SPLASH
  deriving (Show, Eq, Ord)

data Gender
  = GENDER_MALE
  | GENDER_FEMALE
  | GENDER_NEUTER
  deriving (Show, Eq, Ord)

data AnimationType
  = BOTH_DEATH1
  | BOTH_DEAD1
  | BOTH_DEATH2
  | BOTH_DEAD2
  | BOTH_DEATH3
  | BOTH_DEAD3
  | TORSO_GESTURE
  | TORSO_ATTACK
  | TORSO_ATTACK2
  | TORSO_DROP
  | TORSO_RAISE
  | TORSO_STAND
  | TORSO_STAND2
  | LEGS_WALKCR
  | LEGS_WALK
  | LEGS_RUN
  | LEGS_BACK
  | LEGS_SWIM
  | LEGS_JUMP
  | LEGS_LAND
  | LEGS_JUMPB
  | LEGS_LANDB
  | LEGS_IDLE
  | LEGS_IDLECR
  | LEGS_TURN
  | TORSO_GETFLAG
  | TORSO_GUARDBASE
  | TORSO_PATROL
  | TORSO_FOLLOWME
  | TORSO_AFFIRMATIVE
  | TORSO_NEGATIVE
  | LEGS_BACKCR
  | LEGS_BACKWALK
  | FLAG_RUN
  | FLAG_STAND
  | FLAG_STAND2RUN
  deriving (Show, Eq, Ord, Bounded, Enum, Generic)

instance Hashable AnimationType

data Character
  = Character
  { animationMap  :: !(HashMap AnimationType Animation)
  , gender        :: !Gender
  , footStep      :: !FootStep
  , headOffset    :: !(V3 Float)
  , fixedLegs     :: !Bool
  , fixedTorso    :: !Bool
  }
  deriving Show

