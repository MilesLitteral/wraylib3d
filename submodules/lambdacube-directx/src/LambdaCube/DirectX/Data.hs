module LambdaCube.DirectX.Data where

import Control.Applicative
import Control.Monad
import Data.IORef
import Data.List as L
import Data.Maybe
import Foreign 
--import qualified Data.IntMap as IM
import qualified Data.Map as Map
import qualified Data.Set as Set
import qualified Data.Vector as V
import qualified Data.Vector.Storable as SV

--import Control.DeepSeq

import Data.Word
import Codec.Picture
import Codec.Picture.Types

import Graphics.D3D11.Interface --import Graphics.GL.Core33 
import LambdaCube.DirectX.Type
import LambdaCube.DirectX.Util

-- Buffer
disposeBuffer :: ID3D11Buffer -> IO ()
disposeBuffer (Buffer _ bo) = withArray [bo] $ glDeleteBuffers 1

compileBuffer :: [SimpleVertex] -> IO ID3D11Buffer
compileBuffer arrs = do
    let bd = D3D11BufferDesc
        { byteWidth = fromIntegral $ 8 * sizeOf (undefined :: SimpleVertex)
        , usage = D3D11UsageDefault
        , bindFlags = d3d11BindFlags [D3D11BindVertexBuffer]
        , cpuAccessFlags = 0
        , miscFlags = 0
        , structureByteStride = 0 }
    buffer <- createBuffer device bd arrs                  
    iaSetVertexBuffers deviceContext 0 [(buffer, fromIntegral $ sizeOf (undefined :: SimpleVertex), 0)]
    return $! ID3D11Buffer buffer

-- updateBuffer :: ID3D11Buffer -> [(Int,Array)] -> IO ()
-- updateBuffer (Buffer arrDescs bo) arrs = do
--     glBindBuffer GL_ARRAY_BUFFER bo
--     forM arrs $ \(i,Array arrType cnt setter) -> do
--         let ArrayDesc ty len offset size = arrDescs V.! i
--         when (ty == arrType && cnt == len) $
--             setter $! glBufferSubData GL_ARRAY_BUFFER (fromIntegral offset) (fromIntegral size)
--     glBindBuffer GL_ARRAY_BUFFER 0

-- bufferSize :: ID3D11Buffer -> Int
-- bufferSize = V.length . bufArrays

-- arraySize :: ID3D11Buffer -> Int -> Int
-- arraySize buf arrIdx = arrLength $! bufArrays buf V.! arrIdx

-- arrayType :: ID3D11Buffer -> Int -> ArrayType
-- arrayType buf arrIdx = arrType $! bufArrays buf V.! arrIdx

-- Texture
disposeTexture :: TextureData -> IO ()
disposeTexture (TextureData to) = withArray [to] $ glDeleteTextures 1

-- FIXME: Temporary implemenation
uploadTexture2DToGPU :: DynamicImage -> IO TextureData
uploadTexture2DToGPU = uploadTexture2DToGPU' True False True False

uploadTexture2DToGPU' :: ID3D11Device -> DynamicImage -> IO TextureData
uploadTexture2DToGPU' device bitmap' = do
    let bitmap = case bitmap' of
          ImageRGB8   i@(Image w h _) -> bitmap'
          ImageRGBA8  i@(Image w h _) -> bitmap'
          ImageYCbCr8 i@(Image w h _) -> ImageRGB8 $ convertImage i
          di -> ImageRGBA8 $ convertRGBA8 di
        (width,height) = bitmapSize bitmap
            bitmapSize (ImageRGB8  (Image w h _)) = (w,h)
            bitmapSize (ImageRGBA8 (Image w h _)) = (w,h)
            bitmapSize _ = error "unsupported image type :("
            withBitmap (ImageRGB8  (Image w h v)) f = SV.unsafeWith v $ f (w,h) 3 0
            withBitmap (ImageRGBA8 (Image w h v)) f = SV.unsafeWith v $ f (w,h) 4 0
            withBitmap _ _ = error "unsupported image type :("
        td = D3D11Texture2DDesc
                width
                height
                1
                1
                DxgiFormatD24UnormS8Uint
                (DxgiSampleDesc 1 0)
                D3D11UsageDefault
                (d3d11BindFlags []) --D3D11BindDepthStencil
                0
                0
    to <- createTexture2D device td (SV.toList v :: [Word16])
    return $ TextureData to
