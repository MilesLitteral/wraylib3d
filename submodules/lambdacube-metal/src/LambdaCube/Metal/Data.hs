module LambdaCube.Metal.Data where

import Control.Applicative
import Control.Monad
import Data.IORef
import Data.List as L
import Data.Maybe
import Foreign 
--import qualified Data.IntMap as IM
import qualified Data.Map as Map
import qualified Data.Set as Set
import qualified Data.Vector as V
import qualified Data.Vector.Storable as SV

--import Control.DeepSeq

import Data.Word
import Codec.Picture
import Codec.Picture.Types

import LambdaCube.Metal.Type
import LambdaCube.Metal.Util
import Graphics.GL.Core33
--import LambdaCube.Metal.Bindings  --Graphics.GL.Core33
--Utility Functions
calcDesc (offset,setters,descs) (Array arrType cnt setter) =
          let size = cnt * sizeOfArrayType arrType
            in (size + offset, (offset, size,setter):setters, ArrayDesc arrType cnt offset size:descs)

-- Buffer
disposeBuffer :: Buffer -> IO ()
disposeBuffer (Buffer _ bo) = withArray [bo] $ resetCommandsInBuffer 1 

compileBuffer :: MTLDevice -> [Array] -> IO Buffer
compileBuffer dev arrs = do
    let (bufSize,arrSetters,arrDescs) = foldl' calcDesc (0,[],[]) arrs
    bo <- alloca $! \pbo -> makeBuffer dev pbo 1 MTLResourceOptionCPUCacheModeDefault >> peek pbo --glGenBuffers
    forM_ arrSetters $! \(offset,size,setter) -> setter $! makeBuffer dev (fromIntegral offset) (fromIntegral size) MTLResourceOptionCPUCacheModeDefault
    return $! Buffer (V.fromList $! reverse arrDescs) bo

updateBuffer :: MTLDevice -> Buffer -> [(Int,Array)] -> IO ()
updateBuffer (Buffer arrDescs bo) arrs = forM arrs $ 
    \(i,Array arrType cnt setter) -> do
        let ArrayDesc ty len offset size = arrDescs V.! i
        when (ty == arrType && cnt == len) $ 
            setter $! makeBuffer dev (fromIntegral offset) (fromIntegral size) MTLResourceOptionCPUCacheModeDefault

bufferSize :: Buffer -> Int
bufferSize = V.length . bufArrays

arraySize :: Buffer -> Int -> Int
arraySize buf arrIdx = arrLength $! bufArrays buf V.! arrIdx

arrayType :: Buffer -> Int -> ArrayType
arrayType buf arrIdx = arrType $! bufArrays buf V.! arrIdx

-- Texture
disposeTexture :: TextureData -> IO ()
disposeTexture (TextureData to) = withArray [to] $ metalDeleteTextures --glDeleteTextures 1

-- FIXME: Temporary implemenation
uploadTexture2DToGPU :: DynamicImage -> IO TextureData
uploadTexture2DToGPU = uploadTexture2DToGPU' True False True False

uploadTexture2DToGPU' :: Bool -> DynamicImage -> IO TextureData
uploadTexture2DToGPU' isMip bitmap' = do
    let bitmap = case bitmap' of
                    ImageRGB8   i@(Image w h _) -> bitmap'
                    ImageRGBA8  i@(Image w h _) -> bitmap'
                    ImageYCbCr8 i@(Image w h _) -> ImageRGB8 $ convertImage i
                    di -> ImageRGBA8 $ convertRGBA8 di
        (width,height) = bitmapSize bitmap
                         bitmapSize (ImageRGB8  (Image w h _)) = (w,h)
                         bitmapSize (ImageRGBA8 (Image w h _)) = (w,h)
                         bitmapSize _ = error "unsupported image type :("
                         withBitmap (ImageRGB8 (Image w h v))  f = SV.unsafeWith v $ f (w,h) 3 0
                         withBitmap (ImageRGBA8 (Image w h v)) f = SV.unsafeWith v $ f (w,h) 4 0
                         withBitmap _ _ = error "unsupported image type :("
        to <- alloca $! \pto -> newTextureWithDescriptor pto >> MTLTextureDescriptor PixelFormatBGRA8Unorm width height --peek pto
        replaceRegion 0 v $ castPtr ptr
    when isMip $ replaceRegion 1 v $ castPtr ptr
    return $ TextureData to
