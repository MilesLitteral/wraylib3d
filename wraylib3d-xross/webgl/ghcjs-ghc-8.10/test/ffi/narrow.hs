{-# LANGUAGE ForeignFunctionInterface, JavaScriptFFI #-}
{-
  Numeric values returned through the JavaScriptFFI are
  automatically "sanitized": Non-numeric values are
  coerced to numbers and the values are truncated to
  fall within the range of the target type.

  This only happens for lifted return types. If you use
  UnliftedFFITypes to return Int# or Word# directly,
  you need to do the check yourself.
 -}
module Main where

import Data.Int
import Data.Word

main :: IO ()
main = do
  -- unsafe
  t   "jsu_int"      jsu_int
  t   "jsu_int32"    jsu_int32
  t   "jsu_int16"    jsu_int16
  t   "jsu_int8"     jsu_int8
  tio "jsu_intIO"    jsu_intIO
  tio "jsu_int32IO"  jsu_int32IO
  tio "jsu_int16IO"  jsu_int16IO
  tio "jsu_int8IO"   jsu_int8IO
  t   "jsu_word"     jsu_word
  t   "jsu_word32"   jsu_word32
  t   "jsu_word16"   jsu_word16
  t   "jsu_word8"    jsu_word8
  tio "jsu_wordIO"   jsu_wordIO
  tio "jsu_word32IO" jsu_word32IO
  tio "jsu_word16IO" jsu_word16IO
  tio "jsu_word8IO"  jsu_word8IO
  -- safe
  t   "jss_int"      jss_int
  t   "jss_int32"    jss_int32
  t   "jss_int16"    jss_int16
  t   "jss_int8"     jss_int8
  tio "jss_intIO"    jss_intIO
  tio "jss_int32IO"  jss_int32IO
  tio "jss_int16IO"  jss_int16IO
  tio "jss_int8IO"   jss_int8IO
  t   "jss_word"     jss_word
  t   "jss_word32"   jss_word32
  t   "jss_word16"   jss_word16
  t   "jss_word8"    jss_word8
  tio "jss_wordIO"   jss_wordIO
  tio "jss_word32IO" jss_word32IO
  tio "jss_word16IO" jss_word16IO
  tio "jss_word8IO"  jss_word8IO
  -- bool
  b pure "jsu_bone"   jsu_bone
  b pure "jss_bone"   jss_bone
  b id   "jsu_boneIO" jsu_boneIO
  b id   "jss_boneIO" jss_boneIO
  b pure "jsu_bzero"   jsu_bzero
  b pure "jss_bzero"   jss_bzero
  b id   "jsu_bzeroIO" jsu_bzeroIO
  b id   "jss_bzeroIO" jss_bzeroIO

-- 
tio :: (Show a, Integral a) => String -> IO a -> IO ()
tio name x = x >>= \y -> putStrLn (name ++ ": " ++ show y ++ " " ++ show (y == fromInteger c))

t :: (Show a, Integral a) => String -> a -> IO ()
t name x = tio name (pure x)

c :: Integer
c = 123456

b :: (a -> IO Bool) -> String -> a -> IO ()
b f name x = f x >>= \y -> putStrLn (name ++ ": " ++ show y ++ " " ++ show (y == False) ++ " " ++ show (y == True))

-- unsafe
foreign import javascript unsafe "123456.789" jsu_int      :: Int
foreign import javascript unsafe "123456.789" jsu_int32    :: Int32
foreign import javascript unsafe "123456.789" jsu_int16    :: Int16
foreign import javascript unsafe "123456.789" jsu_int8     :: Int8

foreign import javascript unsafe "123456.789" jsu_intIO    :: IO Int
foreign import javascript unsafe "123456.789" jsu_int32IO  :: IO Int32
foreign import javascript unsafe "123456.789" jsu_int16IO  :: IO Int16
foreign import javascript unsafe "123456.789" jsu_int8IO   :: IO Int8

foreign import javascript unsafe "123456.789" jsu_word     :: Word
foreign import javascript unsafe "123456.789" jsu_word32   :: Word32
foreign import javascript unsafe "123456.789" jsu_word16   :: Word16
foreign import javascript unsafe "123456.789" jsu_word8    :: Word8

foreign import javascript unsafe "123456.789" jsu_wordIO   :: IO Int
foreign import javascript unsafe "123456.789" jsu_word32IO :: IO Word32
foreign import javascript unsafe "123456.789" jsu_word16IO :: IO Word16
foreign import javascript unsafe "123456.789" jsu_word8IO  :: IO Word8

-- safe
foreign import javascript safe   "123456.789" jss_int      :: Int
foreign import javascript safe   "123456.789" jss_int32    :: Int32
foreign import javascript safe   "123456.789" jss_int16    :: Int16
foreign import javascript safe   "123456.789" jss_int8     :: Int8

foreign import javascript safe   "123456.789" jss_intIO    :: IO Int
foreign import javascript safe   "123456.789" jss_int32IO  :: IO Int32
foreign import javascript safe   "123456.789" jss_int16IO  :: IO Int16
foreign import javascript safe   "123456.789" jss_int8IO   :: IO Int8

foreign import javascript safe   "123456.789" jss_word     :: Word
foreign import javascript safe   "123456.789" jss_word32   :: Word32
foreign import javascript safe   "123456.789" jss_word16   :: Word16
foreign import javascript safe   "123456.789" jss_word8    :: Word8

foreign import javascript safe   "123456.789" jss_wordIO   :: IO Word 
foreign import javascript safe   "123456.789" jss_word32IO :: IO Word32
foreign import javascript safe   "123456.789" jss_word16IO :: IO Word16
foreign import javascript safe   "123456.789" jss_word8IO  :: IO Word8

-- bool
foreign import javascript unsafe "1" jsu_bone    :: Bool
foreign import javascript safe   "1" jss_bone    :: Bool
foreign import javascript unsafe "1" jsu_boneIO  :: IO Bool
foreign import javascript safe   "1" jss_boneIO  :: IO Bool

foreign import javascript unsafe "0" jsu_bzero    :: Bool
foreign import javascript safe   "0" jss_bzero    :: Bool
foreign import javascript unsafe "0" jsu_bzeroIO  :: IO Bool
foreign import javascript safe   "0" jss_bzeroIO  :: IO Bool

