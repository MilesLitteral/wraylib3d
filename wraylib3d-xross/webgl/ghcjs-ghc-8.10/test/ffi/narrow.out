jsu_int: 123456 True
jsu_int32: 123456 True
jsu_int16: -7616 True
jsu_int8: 64 True
jsu_intIO: 123456 True
jsu_int32IO: 123456 True
jsu_int16IO: -7616 True
jsu_int8IO: 64 True
jsu_word: 123456 True
jsu_word32: 123456 True
jsu_word16: 57920 True
jsu_word8: 64 True
jsu_wordIO: 123456 True
jsu_word32IO: 123456 True
jsu_word16IO: 57920 True
jsu_word8IO: 64 True
jss_int: 123456 True
jss_int32: 123456 True
jss_int16: -7616 True
jss_int8: 64 True
jss_intIO: 123456 True
jss_int32IO: 123456 True
jss_int16IO: -7616 True
jss_int8IO: 64 True
jss_word: 123456 True
jss_word32: 123456 True
jss_word16: 57920 True
jss_word8: 64 True
jss_wordIO: 123456 True
jss_word32IO: 123456 True
jss_word16IO: 57920 True
jss_word8IO: 64 True
jsu_bone: True False True
jss_bone: True False True
jsu_boneIO: True False True
jss_boneIO: True False True
jsu_bzero: False True False
jss_bzero: False True False
jsu_bzeroIO: False True False
jss_bzeroIO: False True False
